std/meta/index.zig

```diff
@@ -0,0 +1,691 @@
+    debug.assert(mem.eql(u8, e2f[0].name, "A"));
+    debug.assert(mem.eql(u8, sf[0].name, "a"));
+    debug.assert(mem.eql(u8, uf[0].name, "a"));
+    debug.assert(comptime sf[0].field_type == u8);
+    debug.assert(comptime uf[0].field_type == u8);
+}
+
+pub fn fieldInfo(comptime T: type, comptime field_name: []const u8) switch (@typeInfo(T)) {
+    TypeId.Struct => TypeInfo.StructField,
+    TypeId.Union => TypeInfo.UnionField,
+    TypeId.ErrorSet => TypeInfo.Error,
+    TypeId.Enum => TypeInfo.EnumField,
+    else => @compileError("Expected struct, union, error set or enum type, found '" 
+        ++ @typeName(T) ++ "'"),
+} {
+    inline for (comptime fields(T)) |field| {
+        if (comptime mem.eql(u8, field.name, field_name))
+            return field;
+    }
+
+    @compileError("'" ++ @typeName(T) ++ "' has no field '" ++ field_name ++ "'");
+}
+
+test "std.meta.fieldInfo" {
+    const E1 = enum.{
+        A,
+    };
+    const E2 = error.{A};
+    const S1 = struct.{
+        a: u8,
+    };
+    const U1 = union.{
+        a: u8,
+    };
+
+    const e1f = comptime fieldInfo(E1, "A");
+    const e2f = comptime fieldInfo(E2, "A");
+    const sf = comptime fieldInfo(S1, "a");
+    const uf = comptime fieldInfo(U1, "a");
+
+    debug.assert(mem.eql(u8, e1f.name, "A"));
+    debug.assert(mem.eql(u8, e2f.name, "A"));
+    debug.assert(mem.eql(u8, sf.name, "a"));
+    debug.assert(mem.eql(u8, uf.name, "a"));
+    debug.assert(comptime sf.field_type == u8);
+    debug.assert(comptime uf.field_type == u8);
+}
+
+pub fn TagType(comptime T: type) type {
+    return switch (@typeInfo(T)) {
+        TypeId.Enum => |info| info.tag_type,
+        TypeId.Union => |info| if(info.tag_type) |Tag| Tag else null,
+        else => @compileError("expected enum or union type, found '" ++ @typeName(T) ++ "'"),
+    };
+}
+
+test "std.meta.TagType" {
+    const E = enum(u8).{
+        C = 33,
+        D,
+    };
+    const U = union(E).{
+        C: u8,
+        D: u16,
+    };
+
+    debug.assert(TagType(E) == u8);
+    debug.assert(TagType(U) == E);
+}
+
+
+
+///Returns the active tag of a tagged union
+pub fn activeTag(u: var) @TagType(UnwrapContainer(@typeOf(u)))
+{
+    const T = @typeOf(u);
+    if(comptime trait.isSingleItemPtr(T)) return @TagType(Child(T))(u.*);
+    return @TagType(T)(u);
+}
+
+test "std.meta.activeTag"
+{
+    const UE = enum.
+    {
+        Int,
+        Float,
+    };
+    
+    const U = union(UE).
+    {
+        Int: u32,
+        Float: f32,
+    };
+    
+    var u = U.{ .Int = 32, };
+    debug.assert(activeTag(u) == UE.Int);
+    
+    u = U.{ .Float = 112.9876, };
+    debug.assert(activeTag(u) == UE.Float);
+
+}
+
+//see: bytesOf
+fn BytesReturnType(comptime P: type) type
+{
+    return if(trait.isConstPtr(P)) []const u8 else []u8;
+}
+///Given a pointer to a single item, returns a slice of the underlying bytes, preserving the
+/// pointer's const-ness.
+pub fn bytes(ptr: var) BytesReturnType(@typeOf(ptr))
```

Comment:
Maybe this should be in `std.mem` instead. We should also probably return an array pointer instead of a slice, as we know the size of the data we are returning. I have a simular implementation of this [here](https://github.com/Hejsil/fun-with-zig/blob/6f5524ef5c8ef1baa32fd253c5fb216dee24104a/src/generic/index.zig#L17-L43). I think we could have `toBytes` and `asBytes` where `asBytes` return a slice with the correct attributes, and `toBytes` just returns a copy of the data as bytes.

Reply:
And maybe we should also have a `bytesToValue` and `bytesAsValue` that takes an array/array pointer respectively and returns a packed value/pointer to packed value.

Reply:
Makes sense to me.